File: 5-kyu/Maximum subarray sum/javascript/Maximum subarray sum.js
/* My Soultion */
const maxSequence = function(arr){
    let curSum = 0;
    return arr.reduce((maxSum, curNum) => {
      curSum = Math.max(curSum + curNum, 0);
      return Math.max(maxSum, curSum);
    }, 0)  
}

/* Best Pratices & Clever */
var maxSequence = function(arr){
    var min = 0, ans = 0, i, sum = 0;
    for (i = 0; i < arr.length; ++i) {
      sum += arr[i];
      min = Math.min(sum, min);
      ans = Math.max(ans, sum - min);
    }
    return ans;
}

File: 5-kyu/Maximum subarray sum/python/Maximum subarray sum.py
# My Soultion using reduce
def max_sequence(arr):
    cur_sum = 0

    def func(max_sum, cur_num):
        nonlocal cur_sum
        cur_sum = max(cur_sum + cur_num, 0)
        return max(max_sum, cur_sum)

    return reduce(func, arr, 0)


# My Soultion using generator
def gen(arr):
    cur_sum = 0
    max_sum = 0

    for cur_num in arr:
        cur_sum = max(cur_sum + cur_num, 0)
        yield max(cur_sum, max_sum)


def max_sequence(arr):
    if not arr:
        return 0

    return max(gen(arr))

# Best Practices & Clever
def maxSequence(arr):
    max,curr=0,0
    for x in arr:
        curr+=x
        if curr<0:curr=0
        if curr>max:max=curr
    return max

# Best Practices & Clever 2
def maxSequence(arr):
  lowest = ans = total = 0
  for i in arr:
    total += i
    lowest = min(lowest, total)
    ans = max(ans, total - lowest)
  return ans